echopype/calibrate/cal_params.py

```diff
@@ -1,155 +1,249 @@
+                out_dict[p_name] = param2da(p_val, channel)
+
+            # p_val has to be one of int, float, xr.DataArray
+            else:
+                raise ValueError(f"{p_name} has to be a scalar, list, or an xr.DataArray")
+
+            # TODO: Consider pre-sort the param xr.DataArray?
 
     return out_dict
 
 
-def get_vend_filter_EK80(
-    vend: xr.Dataset, channel_id: str, filter_name: str, param_type: str
-) -> Union[np.ndarray, int]:
+def _get_interp_da(
+    da_param: Union[None, xr.DataArray],
+    freq_center: xr.DataArray,
+    alternative: Union[int, float, xr.DataArray],
+) -> xr.DataArray:
     """
-    Get filter coefficients stored in the Vendor_specific group attributes.
+    Get interpolated xr.DataArray aligned with the channel coordinate.
+
+    Interpolation at freq_center when da_param contains frequency-dependent xr.DataArray.
+    When da_param is None or does not contain frequency-dependent xr.DataArray,
+    the alternative (a const or an xr.DataArray with coordinate channel) is used.
 
     Parameters
     ----------
-    vend: xr.Dataset
-        EchoData["Vendor_specific"]
-    channel_id : str
-        channel id for which the param to be retrieved
-    filter_name : str
-        name of filter coefficients to retrieve
-    param_type : str
-        'coeff' or 'decimation'
+    da_param : xr.DataArray or None
+        a data array from the Vendor group or user dict with freq-dependent param values
+    freq_center : xr.DataArray
+        center frequency (BB) or nominal frequency (CW)
+    alternative : xr.DataArray or int or float
+        alternative for when freq-dep values do not exist
 
     Returns
     -------
-    np.ndarray or int
-        The filter coefficient or the decimation factor
+    an xr.DataArray aligned with the channel coordinate.
+
+    Note
+    ----
+    ``da_param`` is always an xr.DataArray from the Vendor-specific group.
+    It is possible that only a subset of the channels have frequency-dependent parameter values.
+    The output xr.DataArray here is constructed channel-by-channel to allow for this flexibility.
+
+    ``alternative`` can be one of the following:
+
+    - scalar (int or float): this is the case for impedance_transmit
+    - xr.DataArray with coordinates channel, ping_time, and beam:
+        this is the case for parameters angle_offset_alongship, angle_offset_athwartship,
+                                        beamwidth_alongship, beamwidth_athwartship
+    - xr.DataArray with coordinates channel, ping_time:
+        this is the case for sa_correction and gain_correction,
+        which will be direct output of get_vend_cal_params_power()
     """
-    if param_type == "coeff":
-        v = vend.attrs["%s %s filter_r" % (channel_id, filter_name)] + 1j * np.array(
-            vend.attrs["%s %s filter_i" % (channel_id, filter_name)]
+    param = []
+    for ch_id in freq_center["channel"].data:
```

Comment:
```suggestion
    for ch_id in freq_center["channel"].values:
```

`.data` assumes that it's a numpy array, let's make it so with `.values` so if it happens to be `dask array` you won't have a problem looping here.